packages/langium/src/grammar/validation/validator.ts

```diff
@@ -1013,3 +1119,42 @@ function findLookAheadGroup(rule: AstNode | undefined): ast.TerminalGroup | unde
         return findLookAheadGroup(terminalGroup.$container);
     }
 }
+
+interface AssignmentUse {
+    assignments: Set<ast.Assignment | ast.Action>;
+    /**
+     * Note, that this number is not exact and "estimates the potential number",
+     * i.e. multiplicities like + and * are counted as 2x/twice,
+     * and for alternatives, the worst case is assumed.
+     * In other words, here it is enough to know, whether there are two or more assignments possible to the same feature.
+     */
+    counter: number;
+}
+
+function storeAssignmentUse(map: Map<string, AssignmentUse>, feature: string, increment: number, ...assignments: Array<ast.Assignment | ast.Action>) {
+    let entry = map.get(feature);
+    if (!entry) {
+        entry = {
+            assignments: new Set(),
+            counter: 0,
+        };
+        map.set(feature, entry);
+    }
+    assignments.forEach(a => entry!.assignments.add(a)); // a Set is necessary, since assignments in Fragements might be used multiple times by different parser rules, but they should be marked only once!
+    entry.counter += increment;
+}
+
+function mergeAssignmentUse(mapSoure: Map<string, AssignmentUse>, mapTarget: Map<string, AssignmentUse>, counterOperation: (s: number, t: number) => number = (s, t) => s + t): void {
+    for (const sourceEntry of mapSoure.entries()) {
```

Comment:
Instead of...

> `for (const sourceEntry of mapSoure.entries())`

try in-place deconstructing

`for (const [key, source] of mapSoure.entries())`

Reply:
Thanks!